src/control/ros/RosTrajectoryExecutor.cpp

```diff
@@ -0,0 +1,290 @@
+    waypoint.positions.assign(tangentVector.data(),
+      tangentVector.data() + tangentVector.size());
+
+    assert(0 <= numDerivatives && numDerivatives <= 2);
+    const std::array<std::vector<double>*, 2> derivatives{
+      &waypoint.velocities, &waypoint.accelerations};
+    for (int iderivative = 1; iderivative <= numDerivatives; ++iderivative)
+    {
+      _traj->evaluateDerivative(timeAbsolute, iderivative, tangentVector);
+      assert(tangentVector.size() == numDofs);
+
+      derivatives[iderivative - 1]->assign(tangentVector.data(),
+        tangentVector.data() + tangentVector.size());
+    }
+
+    goal.trajectory.points.emplace_back(waypoint);
+  }
+
+  if (!waitForServer())
+    throw std::runtime_error("Unable to connect to action server.");
+
+  {
+    std::lock_guard<std::mutex> lock(mMutex);
+
+    if (mInProgress)
+      throw std::runtime_error("Another trajectory is in progress.");
+
+    mPromise = std::promise<void>();
+    mInProgress = true;
+    mGoalHandle = mClient.sendGoal(goal, 
+      boost::bind(&RosTrajectoryExecutor::transitionCallback, this, _1));
+
+    return mPromise.get_future();
+  }
+}
+
+//=============================================================================
+bool RosTrajectoryExecutor::waitForServer()
+{
+  using Clock = std::chrono::steady_clock;
+
+  const auto startTime = Clock::now();
+  const auto endTime = startTime + mConnectionTimeout;
+  auto currentTime = startTime + mConnectionPollingRate;
+
+  while (currentTime < endTime)
+  {
+    mCallbackQueue.callAvailable();
+
+    // TODO: Is this thread safe?
+    if (mClient.isServerConnected())
+      return true;
+
+    currentTime += mConnectionPollingRate;
+    std::this_thread::sleep_until(currentTime);
+  }
+
+  return false;
+}
+
+//=============================================================================
+void RosTrajectoryExecutor::transitionCallback(GoalHandle _handle)
+{
+  // This function assumes that mMutex is locked.
+
+  using actionlib::TerminalState;
+  using Result = control_msgs::FollowJointTrajectoryResult;
+
+#if 0
+  ROS_INFO("Transition: CommState = %s TerminalState = %s",
+    _handle.getCommState().toString().c_str(),
+    _handle.getTerminalState().toString().c_str());
+#endif
+  
+  if (_handle.getCommState() == actionlib::CommState::DONE)
+  {
+    std::stringstream message;
+    bool isSuccessful = true;
+
+    // Check the status of the actionlib call. Note that the actionlib call can
+    // succeed, even if execution failed.
+    const auto terminalState = _handle.getTerminalState();
+    if (terminalState != TerminalState::SUCCEEDED)
+    {
+      message << "Action " << terminalState.toString();
+
+      const auto terminalMessage = terminalState.getText();
+      if (!terminalMessage.empty())
+        message << " (" << terminalMessage << ")";
+
+      isSuccessful = false;
+    }
+    else
+    {
+      message << "Execution failed";
```

Comment:
Believe it or not, this is actually correct! 😈 

There are two classes of errors in this function:

1. The state of the [`actionlib` state machine](http://wiki.ros.org/actionlib/DetailedDescription) is reflected by `getCommState()`. If the goal handle finishes _for any reason_ - even if communication is lost with the action server - the state machine transitions to `CommState::DONE`.
2. Once in `CommState::DONE`, `getTerminalState()` indicates why the goal terminated. This can be `TerminalState::SUCCEEDED` or one of several error conditions.
3. If the goal finished in `TerminalState::SUCCEEDED`, then `getResult()` contains the result message. In the case of `FollowJointTrajectoryResult`, the message contains an additional `error_code`.

To summarize: (1) and (2) are **`actionlib` errors** and (3) is a **`FollowJointTrajectory` error**. It's possible for the `actionlib` call to succeed, but for executing the trajectory to fail (in fact: this is the most common case of failure we observe in practice).

The `else` clause here populates `message` with "Execution failed" to differentiate `actionlib` errors from controller errors. Note that `isSuccessful` is **not** set to `false`, so the `message` will only be used if the `result->error_code != Result::SUCCESSFUL` condition fails below.

Does that make sense? Maybe add a comment or two to explain this? 😅